### PR TITLE
TS-4487: fix write_to_net_io issues

### PR DESCRIPTION
**Pull request by oknet**

### CODE REVIEW

File: iocore/net/UnixNetVConnection.cc [hunk 1]
```diff
@@ -545,57 +545,61 @@ write_to_net_io(NetHandler *nh, UnixNetVConnection *vc, EThread *thread)
...
+    if (lock.get_mutex() != s->vio.mutex.get()) {
       write_reschedule(nh, vc);
+      return;
     }
-    if ((needs & EVENTIO_READ) == EVENTIO_READ) {
-      read_reschedule(nh, vc);
-    }
+  }
+
+  if (needs==0 && !buf.reader()->read_avail()) {
```
Comment (SolidWallOfCode) [inline iocore/net/UnixNetVConnection.cc]: Should use `read_avail_more_than(0)` here for performance as the exact number of bytes is irrelevant.

File: iocore/net/UnixNetVConnection.cc [hunk 2]
```diff
@@ -545,57 +545,61 @@ write_to_net_io(NetHandler *nh, UnixNetVConnection *vc, EThread *thread)
...
-    buf.reader()->consume(r);
-    ink_assert(buf.reader()->read_avail() >= 0);
-    s->vio.ndone += r;
+  NET_SUM_DYN_STAT(net_write_bytes_stat, r);
 
-    // If the empty write buffer trap is set, clear it.
-    if (!(buf.reader()->is_read_avail_more_than(0)))
-      vc->write_buffer_empty_event = 0;
+  // Remove data from the buffer and signal continuation.
+  ink_assert(buf.reader()->read_avail() >= r);
```
Comment (SolidWallOfCode) [inline iocore/net/UnixNetVConnection.cc]: Use `is_read_avail_more_than(r)` and `is_read_avail_more_than(0)`.

**Comment by shinrich** [comment]
I think the right thing to do is push PR #629.  That moves the read/write loops to use the IOBuffer interface instead of the IOBlock interface.  I've been holding off on this one until the event loop changes landed since it solved a real problem I had seen.  But if other folks are seeing problems in this area  the IOBuffer interface code is much easier to reason about and debug.

**Comment by oknet** [comment]
vote on PR #629, but the 2nd issue: did not check the change of lock after return from wbe callback, not included.

**Comment by shinrich** [comment]
Good point on the lock check.  Could have finished the current vio in the signal and then end of mysteriously disabling the next vio.

**Comment by oknet** [comment]
Due to the #629 already merged, I will update the patch to fit it.

**Comment by zwoop** [comment]
@oknet This needs a rebase please. @shinrich Can you review please?

**Comment by oknet** [comment]
@zwoop @shinrich a new PR#754 created based on the newest master branch.

**Pull request closed without merging**